### DISCUSSION TITLE
Anisotropic Materials in 3D Printing

### DISCUSSION THREAD

**Discussion by Preetzl** [kinnala/scikit-fem#1125] [Q&A]
Hello kinnala, 

Thank you for creating a very nice library! I was wondering if it is possible to implement a different elasticity module for problems like ex11 as I think that this is what you need to simulate 3d printed parts.

Kind regards

**Comment by kinnala** [comment]
If I understand correctly, you want to have a different elastic modulus for the different parts of the domain.

Yes, it is possible to define the elastic modulus element-by-element. Since I don't know exactly what you have in mind, perhaps you can elaborate it with a drawing?

**Reply by Preetzl** [reply]
![WhatsApp Bild 2024-05-17 um 13 28 07_f7551179](https://github.com/kinnala/scikit-fem/assets/146739548/e2f499e5-c8aa-46af-ae5a-801d1a57112c)

3D-printer is printing each layer on top of each other that we have differnet elastic modulus in z- direction. 
 For my example, I simply apply a force to all nodes on the left side and hold all nodes on the right side (as if the cube was against a wall). 
I would then like to repeat this on all six sides. 
The anisotropy of the material should be taken into account. :) 
A few hurdles, but nothing impossible. 

I'm currently trying to solve the problem of reliably converting stl files into a good mesh without fractals.

**Reply by kinnala** [reply]
There are two options:

1. Specify the material parameters element-by-element
2. Specify the material parameters using inequalities.

This is how you do the latter:

```python

from skfem.helpers import *

@BilinearForm
def weakform(u, v, w):
    mu = 10 * (w.x[2] < 0.5) + 100 * (w.x[2] >= 0.5)
    lmbda = 10 * (w.x[2] < 0.5) + 20 * (w.x[2] >= 0.5)
    epsu = sym_grad(u)
    Cepsu = 2. * mu * epsu + lmbda * eye(trace(epsu), epsu.shape[0])
    return ddot(Cepsu, sym_grad(v))

```

**Reply by Preetzl** [reply]
Thank you for your fast responds. I still couldn`t figure it out to implement the anisotropic elasticity form. I want to try the following steps described in the post: https://en.wikipedia.org/wiki/Transverse_isotropy. 

The stiffness matrix for a simple cube is given there. Since I'm clearly too inexperienced in programming and also don't quite understand how you make your composition of the stiffness matrix, I wanted to ask how you would do it. Is it even possible to keep it that simple?


Sorry for the wrong question at the beginning and the late answer.

**Reply by kinnala** [reply]
I misunderstood your question completely, I thought you want to simply have a different elasticity modulus for the different layers but still consider an isotropic material. However, now I think that what you want to do is to specify a general elasticity 4-tensor $C_{ijkl}$. This should be an easy modification of the variational form. In general, the bilinear form for linear elasticity is
$$\int_\Omega C_{ijkl} \varepsilon_{ij}(u) \varepsilon_{kl}(v) dx.$$
Now we simply need to implement this bilinear form in code.

**Reply by kinnala** [reply]
I don't like the Voigt notation, I see it as an unnecessary mathematical trick to turn the contraction between 4-tensor $C$ and 2-tensor $\varepsilon$ into a matrix-vector product which people might sometimes be more familiar with. However, if you want to use it, you can then do the mapping from and to Voigt notation.

**Reply by kinnala** [reply]
Anyhow, if you want to define a general 4-tensor in numpy, you can do it as follows:
```python
import numpy as np
# initialize 4-tensor
C = np.zeros((3, 3, 3, 3))
# set correct values to all of its components
C[0, 0, 0, 0] = ...
C[0, 0, 0, 1] = ...
```
That should be done according to your anisotropic material model.

**Reply by kinnala** [reply]
The next step is to implement the general bilinear form, I think the following should be enough for all cases:
```python
from skfem import *
from skfem.helpers import *

# assuming C is defined as above ...

@BilinearForm
def anisotropic(u, v, _):
    epsu = 0.5 * (grad(u) + transpose(grad(u)))
    epsv = 0.5 * (grad(v) + transpose(grad(v)))
    return np.einsum('ijkl,ij...,kl...', C, epsu, epsv)
```
